File: Filewatcherd-Go/src/codewind/httpgetstatusthread.go
/*******************************************************************************
* Copyright (c) 2019 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

package main

import (
	"codewind/models"
	"codewind/utils"
	"crypto/tls"
	"encoding/json"
	"errors"
	"io/ioutil"
	"net/http"
	"strconv"
	"strings"
	"time"
)

/**
 * This file is responsible for issuing a GET request to the server in order to
 * retrieve the latest list of projects to watch (including their path, and any
 * filters).
 *
 * A new GET request will be sent by this class on startup, and after startup, a
 * GET request will be sent either: whenever the WebSocket connection fails, or
 * otherwise, once every X seconds (eg 120).
 *
 * ws.go is responsible for informing this code when the WebSocket
 * connection fails (input), and this class calls the Filewatcher
 * class with the data from the GET request (containing any project watch
 * updates received) as output.
 *
 */
type HttpGetStatusThread struct {
	refreshStatusChan chan interface{}
	baseURL           string
}

/**
 * This function is called by other files whenever a new GET request should be sent to the server (for example, if
 * the websocket connecion failed.) */
func (hg *HttpGetStatusThread) SignalStatusRefreshNeeded() {
	utils.LogDebug("SignalStatusRefreshNeeded called.")
	hg.refreshStatusChan <- nil
	utils.LogDebug("post SignalStatusRefreshNeeded called.")
}

func NewHttpGetStatusThread(baseURL string, projectList *ProjectList) (*HttpGetStatusThread, error) {

	baseURL = utils.StripTrailingForwardSlash(baseURL)

	if !utils.IsValidURLBase(baseURL) {
		return nil, errors.New("URL is invalid: " + baseURL)
	}

	reconnectNeeded := make(chan interface{})

	result := &HttpGetStatusThread{
		reconnectNeeded,
		baseURL,
	}

	go runGetStatusThread(result, projectList)

	result.SignalStatusRefreshNeeded()

	// Every 120 seconds, refresh the status
	ticker := time.NewTicker(120 * time.Second)
	go func() {
		for {
			<-ticker.C
			utils.LogDebug("GetStatus ticker ticked.")
			result.SignalStatusRefreshNeeded()
		}
	}()

	return result, nil

}

func runGetStatusThread(data *HttpGetStatusThread, projectList *ProjectList) {
	utils.LogInfo("Http GET status thread started.")

	backoff := utils.NewExponentialBackoff()

	for {
		// Wait for at least one request
		<-data.refreshStatusChan

		// Once a refresh status request is issued, keep trying until it succeeds.
		success := false
		for !success {

			err := doGetRequest(data.baseURL, backoff.GetFailureDelay(), projectList)
			if err != nil {
				utils.LogErrorErr("Error from GET request", err)
				backoff.SleepAfterFail()
				backoff.FailIncrease()
			} else {
				backoff.SuccessReset()
				success = true
			}
		}

		// On success, drain the channel of any other requests that occurred during this time.
		channelEmpty := false
		for !channelEmpty {
			select {
			case <-data.refreshStatusChan:
			default:
				channelEmpty = true
			}
		}

		utils.LogDebug("GET request successfully sent and received.")

	} // end for
}

func doGetRequest(baseURL string, failureDelay int, projectList *ProjectList) error {

	// Wait before issuing a request, due to a previous failed request
	if failureDelay > 0 {
		time.Sleep(time.Duration(failureDelay) * time.Millisecond)
	}
	result, err := sendGet(baseURL)

	if err != nil {
		return err
	}

	if result != nil {
		projectList.UpdateProjectListFromGetRequest(result)
	}

	return nil

}

func sendGet(baseURL string) (*models.WatchlistEntries, error) {

	url := baseURL + "/api/v1/projects/watchlist"

	utils.LogInfo("Initiating GET request to " + url)

	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}

	client := &http.Client{Transport: tr}

	resp, err := client.Get(url)
	if err != nil || resp == nil {
		errMsg := "Get request failed for " + url + " , with no response code."
		if err != nil {
			utils.LogErrorErr(errMsg, err)
		} else {
			utils.LogError(errMsg)
		}

		return nil, err
	}

	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		errMsg := "Get response failed for " + url + ", response code: " + strconv.Itoa(resp.StatusCode)
		utils.LogError(errMsg)
		return nil, errors.New(errMsg)
	}

	body, err := ioutil.ReadAll(resp.Body)

	if err != nil || body == nil {
		utils.LogError("Get response failed for " + url + ", unable to read body")
		return nil, err
	}

	// Strip EOL characters to ensure it fits on one log line.
	bodyStr := string(body)
	bodyStr = strings.ReplaceAll(bodyStr, "\r", "")
	bodyStr = strings.ReplaceAll(bodyStr, "\n", "")

	utils.LogInfo("GET request completed, for " + url + ". Response: " + bodyStr)

	var entries models.WatchlistEntryList
	err = json.Unmarshal(body, &entries)
	if err != nil {
		utils.LogError("Get response failed for" + url + ", unable to unmarshal body.")
		return nil, err
	}

	return &entries.Projects, nil
}

File: Filewatcherd-TypeScript/src/lib/HttpGetStatusThread.ts
/*******************************************************************************
* Copyright (c) 2019, 2020 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

import * as log from "./Logger";

import got from "got";

import { ExponentialBackoffUtil } from "./ExponentialBackoffUtil";
import { FileWatcher } from "./FileWatcher";
import { ProjectToWatch } from "./ProjectToWatch";

/**
 * This class is responsible for issuing a GET request to the server in order to
 * retrieve the latest list of projects to watch (including their path, and any
 * filters).
 *
 * A new GET request will be sent by this class on startup, and after startup,
 * a GET request will be sent either:
 * - whenever the WebSocket connection fails
 * - or otherwise, once every X seconds (eg 120)
 *
 * WebSocketManagerThread is responsible for informing this class when the
 * WebSocket connection fails (input), and this class calls the Filewatcher
 * class with the data from the GET request (containing any project watch
 * updates received) as output.
 */
export class HttpGetStatusThread {

    public static readonly REFRESH_EVERY_X_SECONDS = 120;

    private _baseUrl: string;

    private _inInnerLoop: boolean;

    private _timer: NodeJS.Timer = null;

    private _disposed: boolean = false;

    private readonly _parent: FileWatcher;

    public constructor(baseUrl: string, parent: FileWatcher) {
        this._baseUrl = baseUrl;
        this._parent = parent;

        this.resetTimer();
        this.queueStatusUpdate();
    }

    public async queueStatusUpdate(): Promise<void> {
        if (this._disposed) { return; }
        if (this._inInnerLoop) { return; }

        this.innerLoop();

    }
    public dispose() {
        if (this._disposed) { return; }

        log.info("dispose() called on HttpGetStatusThread");

        this._disposed = true;
    }

    private async doHttpGet(): Promise<ProjectToWatch[]> {

        const requestObj = {
            header: {},
            rejectUnauthorized: false,
            retry: 0,
            timeout: 20000,
        };

        const authTokenWrapper = this._parent.authTokenWrapper;
        const authToken = authTokenWrapper.getLatestToken();
        if (authToken && authToken.accessToken) {
            requestObj.header = { bearer: authToken.accessToken };
        }

        try {

            const url = this._baseUrl + "/api/v1/projects/watchlist";

            log.info("Initiating GET request to " + url);

            const response = await got(url, requestObj);

            if (response.statusCode  === 200 && response.body) {

                // Strip EOL characters to ensure it fits on one log line.
                const w = JSON.parse(response.body);

                const debugVal = JSON.stringify(w).replace("\n", "").replace("\r", "");

                log.info("GET response received: " + debugVal);

                if (w == null || w === undefined) {
                    log.error("Expected value not found for GET watchlist endpoint");
                    return null;
                }

                const result = new Array<ProjectToWatch>();

                for (const e of w.projects) {

                    // Sanity check the JSON parsing
                    if (!e.projectID || !e.pathToMonitor) {
                        log.error("JSON parsing of GET watchlist endpoint failed with missing values");
                        return null;
                    }

                    result.push(ProjectToWatch.createFromJson(e, false));
                }

                return result;

            } else {

                if (response) {
                    log.info("GET request failed, details: " + response.statusCode + " " + response.body);
                }

                // Inform bad token if we are redirected to an OIDC endpoint
                if (authToken && authToken.accessToken && response.statusCode && response.statusCode === 302
                    && response.headers && response.headers.location
                    && response.headers.location.indexOf("openid-connect/auth") !== -1) {

                    authTokenWrapper.informBadToken(authToken);
                }

                return null;
            }

        } catch (err) {
            log.error("GET request failed. [" + (err.message) + "] (" + err.statusCode + ")");

            // Inform bad token if we are redirected to an OIDC endpoint
            if (err.statusCode === 302 && err.response && err.response.headers && err.response.headers.location
                && err.response.headers.location.indexOf("openid-connect/auth") !== -1) {

                if (authToken && authToken.accessToken) {
                    authTokenWrapper.informBadToken(authToken);
                }

            }

            return null;
        }

    }
    private async innerLoop(): Promise<void> {
        try {
            this._inInnerLoop = true;
            log.debug("HttpGetStatus - beginning GET request loop.");

            let success = false;

            const delay = ExponentialBackoffUtil.getDefaultBackoffUtil(4000);

            let result: ProjectToWatch[] = null;

            while (!success && !this._disposed) {

                result = await this.doHttpGet();

                success = result != null;

                if (!success) {
                    log.error("HTTP get request failed");
                    await delay.sleepAsync();
                    delay.failIncrease();
                }

            }

            if (result && result.length > 0) {
                this._parent.updateFileWatchStateFromGetRequest(result);
            }

        } finally {
            log.info("HttpGetStatus - GET request loop complete.");
            this._inInnerLoop = false;
            this.resetTimer();
        }

    }
    private resetTimer() {
        if (this._timer != null) {
            clearTimeout(this._timer);
        }

        // Refresh every X seconds (eg 120)
        this._timer = setTimeout(() => {
            this.queueStatusUpdate();
        }, HttpGetStatusThread.REFRESH_EVERY_X_SECONDS * 1000);

    }

}
